### ISSUE TITLE
False negatives for missing_docs_in_private_items since 1.79.

### ISSUE THREAD

**Issue by nmathewson** [rust-lang/rust-clippy#14301]
### Summary

Apparently there are circumstances where missing_docs_in_private_items fails to report errors; they seem to have strange, non-local behavior.

### Lint Name

missing_docs_in_private_items

### Reproducer

I ran `cargo clippy --all-features` with Rust 1.85 using this code as my lib.rs.

```rust
//! Documented.

#![deny(clippy::missing_docs_in_private_items)]

/// Documented.
#[derive(Debug)]
#[allow(dead_code)]
struct Mine {
    //// Note use of four /s.  No warning here.
    undocumented_1: String,
}

#[derive(serde::Deserialize)]
pub(crate) struct Undocumented2 {}
```

Here's the Cargo.toml I used:
```rust
[package]
name = "clippy-repro"
version = "0.1.0"
edition = "2021"

[features]
f = []

[dependencies]
serde = { version = "1.0.218", features = ["derive"] }
```

I expected to see warnings from `undocumented_1` and `Undocumented2`.

Instead, no warnings were produced.

I tried removing `#[derive(Debug)]` from `Mine`: It caused a warning to start appearing on `Undocumented2`!

I tried with older versions of Clippy: Rust version 1.78 and earlier correctly give warnings; Rust 1.79 is the first version that fails to warn at all.

### Version

```text
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: x86_64-unknown-linux-gnu
release: 1.85.0
LLVM version: 19.1.7
```

**Comment by nmathewson** [comment]
This might be related to #13996

**Comment by nmathewson** [comment]
I have experimentally tried reverting several of the docs-related commits added between Rust 1.78 and Rust 1.79.  If I revert adc91e4913875e430d946852e28f8a18e40afdfb, I think I see these warnings again.

(I say "I think" because I am not super familiar with hacking on the clippy toolchain.)

**Comment by clarfonthey** [comment]
I seem to be able to reliably trigger this when calling `cargo clippy --tests`: essentially, an `#![expect(clippy::missing_docs_in_private_items)]` will do fine when just running `cargo clippy`, but it seems like there might be a case where it's being run twice for tests, and not triggering the second time? Or something like that?

Hopefully this helps nail down what's happening here, because I don't get it, either.
